### PR TITLE
Laboratory work #1, Maria Kramkova - 22FPL1

### PR DESCRIPTION
**Pull request by kramkova**

### CODE REVIEW

File: lab_1_classify_by_unigrams/main.py [hunk 1]
```diff
@@ -5,6 +5,14 @@
 
 
 def tokenize(text: str) -> list[str] | None:
+    if type(text) != str:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:8]: oh no, use `isinstance`

File: lab_1_classify_by_unigrams/main.py [hunk 2]
```diff
@@ -14,6 +22,17 @@ def tokenize(text: str) -> list[str] | None:
 
 
 def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
+    if type(tokens) != list or tokens == [] or not all(isinstance(element, str) for element in tokens):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:25]: `isinstance` is all you need
if you want to check a container for emptity - use len or BETTER us truthy/falsy : https://stackoverflow.com/questions/39983695/what-is-truthy-and-falsy-how-is-it-different-from-true-and-false

File: lab_1_classify_by_unigrams/main.py [hunk 3]
```diff
@@ -14,6 +22,17 @@ def tokenize(text: str) -> list[str] | None:
...
+        return None
+    frequencies = {}
+    unit = 1/len(tokens)
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:28]: add spapce around `/`

File: lab_1_classify_by_unigrams/start.py [hunk 4]
```diff
@@ -2,6 +2,7 @@
 Language detection starter
 """
 
+#line to be deleted
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/start.py:5]: so, delete)

File: main.py [hunk 5]
```diff
@@ -0,0 +1 @@
+print('Hello, world!')
```
Comment (artyomtugaryov) [inline main.py:1]: and this file delete too

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
I believe you can get a higher score!

File: lab_1_classify_by_unigrams/main.py (hunk 1 above)
Comment (kramkova) [reply lab_1_classify_by_unigrams/main.py]: nice, implemented

File: lab_1_classify_by_unigrams/start.py (hunk 4 above)
Comment (kramkova) [reply lab_1_classify_by_unigrams/start.py]: fixed

File: main.py (hunk 5 above)
Comment (kramkova) [reply main.py]: fixed

File: lab_1_classify_by_unigrams/main.py (hunk 3 above)
Comment (kramkova) [reply lab_1_classify_by_unigrams/main.py]: fixed

File: lab_1_classify_by_unigrams/main.py [hunk 6]
```diff
@@ -6,49 +6,94 @@
 
 def tokenize(text: str) -> list[str] | None:
     """
-    Splits a text into tokens, converts the tokens into lowercase,
-    removes punctuation, digits and other symbols
-    :param text: a text
-    :return: a list of lower-cased tokens without punctuation
-    """
+        Splits a text into tokens, converts the tokens into lowercase,
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:9]: revert changes in all doc strings, please

File: seminars/practice_2_string.py [hunk 7]
```diff
@@ -53,13 +53,14 @@ def multiply_string(input_string: str, how_many: int) -> str:
     display the given string the number of times given in the `how_many`.
     """
     # student realisation goes here
+    return input_string * how_many
```
Comment (artyomtugaryov) [inline seminars/practice_2_string.py:56]: please, revert the changes

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

**Comment by artyomtugaryov** [comment]
The same comments

File: seminars/practice_2_string.py (hunk 7 above)
Comment (kramkova) [reply seminars/practice_2_string.py]: backed up the file

File: lab_1_classify_by_unigrams/main.py (hunk 6 above)
Comment (artyomtugaryov) [reply lab_1_classify_by_unigrams/main.py:9]: !

File: lab_1_classify_by_unigrams/main.py [hunk 8]
```diff
@@ -6,55 +6,103 @@
...
+        removes punctuation, digits and other symbols
+        :param text: a text
+        :return: a list of lower-cased tokens without punctuation
+        """
+    if not isinstance(text, str):
+        return None
+    tokens = []
+    for symbol in text:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:17]: try to use list comprehension

File: lab_1_classify_by_unigrams/main.py [hunk 9]
```diff
@@ -6,55 +6,103 @@
...
+        :return: a dictionary with frequencies
+        """
+    if (not isinstance(tokens, list)
+            or not tokens
+            or not all(isinstance(element, str) for element in tokens)):
+        return None
+    frequencies = {}
+    unit = 1 / len(tokens)
+    for letter in tokens:
+        if letter in frequencies:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:36]: What if we set 0 if `letter` is not in `frequencies` and always do += ...

File: lab_1_classify_by_unigrams/main.py [hunk 10]
```diff
@@ -6,55 +6,103 @@
...
 
 
 def create_language_profile(language: str, text: str) -> dict[str, str | dict[str, float]] | None:
     """
-    Creates a language profile
-    :param language: a language
-    :param text: a text
-    :return: a dictionary with two keys – name, freq
-    """
+        Creates a language profile
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:45]: revert

File: lab_1_classify_by_unigrams/main.py [hunk 11]
```diff
@@ -6,55 +6,103 @@
...
+        """
+    target_set = {'freq', 'name'}
+    if (not isinstance(unknown_profile, dict)
+            or not isinstance(profile_to_compare, dict)
+            or len(target_set & set(unknown_profile) & set(profile_to_compare)) != len(target_set)):
+        return None
+    new_freq = {a: .0 for a in set(list(unknown_profile['freq']) +
+                                   list(profile_to_compare['freq']))}
+    freq_to_compare = new_freq.copy()
+    assert isinstance(unknown_profile['freq'], dict)
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:94]: No! you must return `None`

File: lab_1_classify_by_unigrams/main.py [hunk 12]
```diff
@@ -6,55 +6,103 @@
...
+    new_freq.update(unknown_profile['freq'])
+    assert isinstance(profile_to_compare['freq'], dict)
+    freq_to_compare.update(profile_to_compare['freq'])
+    distance = calculate_mse(list(new_freq.values()), list(freq_to_compare.values()))
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:98]: do not create the variable - return `dict` as is

File: lab_1_classify_by_unigrams/start.py [hunk 13]
```diff
@@ -13,8 +14,15 @@ def main() -> None:
         de_text = file_to_read_de.read()
     with open("assets/texts/unknown.txt", "r", encoding="utf-8") as file_to_read_unk:
         unknown_text = file_to_read_unk.read()
-    result = None
+    unknown_profile = create_language_profile('Unknown', unknown_text)
+    en_profile = create_language_profile('English', en_text)
+    de_profile = create_language_profile('Deutsch', de_text)
+    assert unknown_profile
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/start.py:20]: .

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_1_classify_by_unigrams/main.py (hunk 8 above)
Comment (kramkova) [reply lab_1_classify_by_unigrams/main.py:17]: changed

File: lab_1_classify_by_unigrams/main.py (hunk 6 above)
Comment (kramkova) [reply lab_1_classify_by_unigrams/main.py:9]: reverted the excessive tabulation

File: lab_1_classify_by_unigrams/main.py (hunk 11 above)
Comment (kramkova) [reply lab_1_classify_by_unigrams/main.py:94]: Added it to the if-conditions block

File: lab_1_classify_by_unigrams/main.py (hunk 12 above)
Comment (kramkova) [reply lab_1_classify_by_unigrams/main.py:98]: omitted the 'distance' variable

File: lab_1_classify_by_unigrams/main.py (hunk 10 above)
Comment (kramkova) [reply lab_1_classify_by_unigrams/main.py]: perhaps done

File: lab_1_classify_by_unigrams/main.py (hunk 9 above)
Comment (kramkova) [reply lab_1_classify_by_unigrams/main.py:33]: Thank you, implemented
